### ISSUE TITLE
Upstream: clarify use of spcm->mutex

### ISSUE THREAD

**Issue by plbossart** [thesofproject/linux#654]
feedback from Takashi Iwai

````
> +static int sof_pcm_close(struct snd_pcm_substream *substream)
> +{
> +	struct snd_soc_pcm_runtime *rtd = substream->private_data;
> +	struct snd_soc_component *component =
> +		snd_soc_rtdcom_lookup(rtd, DRV_NAME);
> +	struct snd_sof_dev *sdev =
> +		snd_soc_component_get_drvdata(component);
> +	struct snd_sof_pcm *spcm = rtd->private;
> +	int err;
> +
> +	/* nothing todo for BE */
> +	if (rtd->dai_link->no_pcm)
> +		return 0;
> +
> +	dev_dbg(sdev->dev, "pcm: close stream %d dir %d\n", spcm->pcm.pcm_id,
> +		substream->stream);
> +
> +	snd_sof_pcm_platform_close(sdev, substream);

Doesn't it need spcm->mutex lock while open is called inside mutex?
Or, better to reconsider: what does spcm->mutex protect?

> +
> +	mutex_lock(&spcm->mutex);
> +	pm_runtime_mark_last_busy(sdev->dev);
> +
> +	err = pm_runtime_put_autosuspend(sdev->dev);
> +	if (err < 0)
> +		dev_err(sdev->dev, "error: pcm close failed to idle %d\n",
> +			err);
> +
> +	mutex_unlock(&spcm->mutex);
> +	return 0;
````

Indeed the position of the mutex_lock makes no sense. Should be either before the close or removed completely.

**Comment by lyakh** [comment]
@plbossart TBH I don't understand at all what that mutex is protecting. E.g. in sof_compressed_open(), sof_compressed_free(). Is it trying to do something like (pci_)dev_get() / put()?

**Comment by plbossart** [comment]
@lyakh let's part the compress.c code for now, it's not functional.

@ranj063 @keyonjie @lgirdwood can you clarify what the intent for the spcm->muted was? what does it protect? What is the point of serializing the calls to pm_runtime functions?

**Comment by plbossart** [comment]
it's actually worse than I thought, we have additional mutexes that are never used.
Suggested fix for https://github.com/thesofproject/linux/pull/703

**Comment by lyakh** [comment]
resolved by #703
